### ISSUE TITLE
Timedistributed CNN + LSTM

### ISSUE THREAD

**Issue by muzixiang** [fchollet/keras#3032]
Hi, everyone, I am a newbie in keras, I want to do sequence (each element in the sequence is a 2d matrix or called image) classification (2 classes) just like the first architecture in fig 3 in this article http://arxiv.org/pdf/1411.4389v4.pdf , the author said the label are determined by averaging the softmax of each time steps, this confuse me, can anyone tell me how to average the softmax of each time steps? and return sequences should to be set False or True?? Very thanks!
is my following code is right?

```
    model = Sequential()
    model.add(TimeDistributed(Convolution2D(8, 5, 5, border_mode='valid'), input_shape=(maxToAdd, 1, size1, size2)))
    model.add(TimeDistributed(MaxPooling2D(pool_size=(2, 2),border_mode='valid')))
    model.add(Activation('relu'))
    model.add(TimeDistributed(Flatten()))
    model.add(LSTM(output_dim=64, return_sequences=True))
    model.add(Dropout(.5))
    model.add(TimeDistributed(Dense(1)))
    model.add(Activation('sigmoid'))
    rmsprop = RMSprop()
    model.compile(loss='mean_squared_error', optimizer=rmsprop, metrics=['accuracy'])
```

**Comment by muzixiang** [comment]
@carlthome can you tell me how to implement this?

**Comment by ymcui** [comment]
if that is a 2-class classification, 
as you said `average the softmax of each time steps`, 
this perhaps means to use two neurons in the output layer to represent each class respectively.
here is a sample code AFTER `LSTM` layer,
```
import theano.backend as K
def td_avg(x):
    return K.mean(x, axis=1)
model.add(TimeDistributed(Dense(2,activation='softmax'))
model.add(Lambda(td_avg))
```

**Comment by muzixiang** [comment]
Thanks for your rapid response, can you further tell me why to write like this, I find little official document about lambda, and I don't know what the x in your code is, it represents each timedistributed dense's output?

**Comment by shashankg7** [comment]
@muzixiang As far as my understanding of your model goes model.add(TimeDistributed(Dense(2,activation='softmax')), after this step a 2 dimensional layer will be present and since you want average of the activation of these layers, you need to add a 'mean' layer. Hence the Lambda layer.

**Comment by ymcui** [comment]
@muzixiang i think you'd better take a look at the usage of `Lambda` Layer in official documentation. Then you will know the what is `x` all about. :)

**Comment by muzixiang** [comment]
@ymcui hi I tried your method, but error appears
Exception: Error when checking model target: expected lambda_1 to have 3 dimensions, but got array with shape (8000L, 2L)

8000 is the size of my training set. How to correct this, thanks!

**Comment by ymcui** [comment]
```
import theano.backend as K
def td_avg(x):
    return K.mean(x, axis=1)
def td_avg_shape(x):
    return tuple((batch_size,hidden_dimension))
model.add(TimeDistributed(Dense(2,activation='softmax'))
model.add(Lambda(td_avg,output_shape=td_avg_shape))
```
what if you try this one?
note that, replace `batch_size` and `hidden_dimension` with your true value.

**Comment by muzixiang** [comment]
@ymcui 
Hi. I haven't try your new code, but I found a way that work well, as follows:

time_distributed_merge_layer = Lambda(function=lambda x: K.mean(x, axis=1),
                                      output_shape=lambda shape: (shape[0],) + shape[2:])
model.add(LSTM(output_dim=128, return_sequences=True))
model.add(Dropout(0.5))
model.add(TimeDistributed(Dense(256)))
model.add(Dropout(0.5))
model.add(TimeDistributed(Dense(2, activation='softmax')))
model.add(time_distributed_merge_layer)

**Comment by ymcui** [comment]
neat. that's exactly the same with mine. :)

**Comment by muzixiang** [comment]
@ymcui 
How to save each time step's output, does it also need to be realized in the lambda function?

**Comment by ymcui** [comment]
If you want to get an intermediate output of specific layer,
just write a `theano.function` and indicate the `input` and `output`.
here is an official document: https://keras.io/getting-started/faq/#how-can-i-visualize-the-output-of-an-intermediate-layer

**Comment by muzixiang** [comment]
@ymcui 
Very thanks, another question confuse me is 
K.mean(x, axis=1)
the 0 axis indicate what dimension, and the 1 axis indicate what?

**Comment by shashankg7** [comment]
@muzixiang .. If you have a tensor x with dimensions (d1, d2 .... dk), K.mean(x, axis=i) will take mean along dimension di .. and x will now have dimensions (d1, d2 ...., di-1 , di +1, .... dk)

**Comment by ymcui** [comment]
@muzixiang 
if you are familiar with definition of tensor shape, then it will be easy to understand.
`axis` begins from `0`, so `axis=1` is the second dimension in tensor shape.
For e.g.
if `x` has a shape of `(batch_size, max_length, hidden_size)`, then the `axis=1` means the `max_length` dimension.

**Comment by muzixiang** [comment]
@ymcui 
Sincerely thanks for your rapid answers !!

**Comment by Mako18** [comment]
@muzixiang I having problems creating a network first with CNN layer and then a LSTM layer because of input shape I think. Could you please tell me how did you reshape your input? I also try with Convolution2D with 1730 gray scale images as inputs.

**Comment by Auroras-Li** [comment]
> @muzixiang I having problems creating a network first with CNN layer and then a LSTM layer because of input shape I think. Could you please tell me how did you reshape your input? I also try with Convolution2D with 1730 gray scale images as inputs.

Can you tell me how you solved the input problem? I also want to use CNN with LSTM, but I don't know what to do with the shape of the input.
